llvm/lib/Target/AMDGPU/AMDGPUImageIntrinsicOptimizer.cpp

```diff
@@ -0,0 +1,336 @@
+    const APInt &NewFragIdVal = FragId->getValue().udiv(4) * 4;
+
+    // Create the new instructions.
+    IRBuilder<> B(IIList.front());
+
+    // Create the new image_msaa_load intrinsic.
+    SmallVector<Instruction *, 4> NewCalls;
+    while (DMaskVal != 0) {
+      unsigned NewMaskVal = 1 << countr_zero(DMaskVal);
+
+      Intrinsic::ID NewIntrinID;
+      if (IntrinID == Intrinsic::amdgcn_image_load_2dmsaa)
+        NewIntrinID = Intrinsic::amdgcn_image_msaa_load_2dmsaa;
+      else
+        NewIntrinID = Intrinsic::amdgcn_image_msaa_load_2darraymsaa;
+
+      Function *NewIntrin = Intrinsic::getDeclaration(
+          IIList.front()->getModule(), NewIntrinID, OverloadTys);
+      Args[ImageDimIntr->DMaskIndex] =
+          ConstantInt::get(DMask->getType(), NewMaskVal);
+      Args[FragIdIndex] = ConstantInt::get(FragId->getType(), NewFragIdVal);
+      CallInst *NewCall = B.CreateCall(NewIntrin, Args);
+      LLVM_DEBUG(dbgs() << "Optimize: " << *NewCall << "\n");
+
+      NewCalls.push_back(NewCall);
+      DMaskVal -= NewMaskVal;
+    }
+
+    // Create the new extractelement instructions.
+    for (auto &II : IIList) {
+      Value *VecOp = UndefValue::get(II->getType());
+      auto Idx = cast<ConstantInt>(II->getArgOperand(FragIdIndex));
+      if (NumElts == 1) {
+        VecOp = B.CreateExtractElement(NewCalls[0], Idx->getValue().urem(4));
+        LLVM_DEBUG(dbgs() << "Add: " << *VecOp << "\n");
+      } else {
+        for (unsigned I = 0; I < NumElts; ++I) {
+          VecOp = B.CreateInsertElement(
+              VecOp, B.CreateExtractElement(
+                NewCalls[I], Idx->getValue().urem(4)), I);
+          LLVM_DEBUG(dbgs() << "Add: " << *VecOp << "\n");
+        }
+      }
+
+      // Replace the old instruction.
+      II->replaceAllUsesWith(VecOp);
+      InstrsToErase.push_back(II);
+    }
+
+    Modified = true;
+  }
+
+  for (auto I : InstrsToErase) {
+    I->eraseFromParent();
+  }
+
+  MergeableInsts.clear();
+
+  return Modified;
+}
+
+static bool imageIntrinsicOptimizerImpl(Function &F, const TargetMachine *TM) {
+  if (!TM)
+    return false;
+
+  // This optimization only applies to GFX11 and beyond.
+  const GCNSubtarget &ST = TM->getSubtarget<GCNSubtarget>(F);
+  if (!AMDGPU::isGFX11Plus(ST) || ST.hasMSAALoadDstSelBug())
+    return false;
+
+  Module *M = F.getParent();
+
+  // Early test to determine if the intrinsics are used.
+  if (std::none_of(M->begin(), M->end(), [](Function &F) {
+        return !F.users().empty() &&
+               (F.getIntrinsicID() == Intrinsic::amdgcn_image_load_2dmsaa ||
+                F.getIntrinsicID() == Intrinsic::amdgcn_image_load_2darraymsaa);
+      }))
+    return false;
+
+  bool Modified = false;
+  for (auto &BB : F) {
+    BasicBlock::iterator SectionEnd;
+    for (BasicBlock::iterator I = BB.begin(), E = BB.end(); I != E;
+         I = SectionEnd) {
+      std::list<std::list<IntrinsicInst *>> MergeableInsts;
```

Comment:
Is there any particular reason to use an `std::list<std::list<`?

Using `std::list` is particularly discouraged in the LLVM programmer's manual (https://llvm.org/docs/ProgrammersManual.html#list) unless big objects are involved or stability of the iterators after inserting new elements is needed.

I'd rather use `SmallVector<SmallVector<IntrinsicInst*, 0>, 0>` or `std::vector<std::vector<IntrinsicInst*>>` instead (slight preference for the first since it can use realloc).

Reply:
I switched to SmallVector. Do you think the `, 0` part is necessary? In particular it seems odd to put `, 0` on the outer vector.

Reply:
I guess it depends on the workload. If the vector often goes over the stack space allocated by a `SmallVector<T>` I would use the `, 0` version.

In this case, I think specifying the `, 0` or not will be fine in either case.

Just to keep in mind, on x86_64:
* a `SmallVector<void*>` preallocates stack space for 6 elements
* a `SmallVector<SmallVector<void*>>` prealocates stack space for a single element
* a `SmallVector<SmallVector<void*, 0>>` prealocates stack space for 3 elements

Reply:
I used `, 4` for the inner list. It should be very unusual for it to get longer than that, because of the way the lists contain instructions that are identical except for having different FragIds in the range 0..3.